# Cryptographie — Travail pratique 2

**Auteurs**
- Jeddi Youssef
- Themlaoui Houssem

**Lien repo Github**  : 
https://github.com/ThemlaouiHou/travail-pratique-1-




## Exercice 1 — Accès GitLab (SSH)

**À compléter**
- Algorithme de chiffrement: ...
- Longueur de clé: ...
- Clé publique: ...

**Sécurité (attaques classiques)**
- ...

**Sécurité (ordinateur quantique)**
- ...


## Exercice 2 — Devoir 7, exercice 6 (hachage)

On considère la fonction de hachage:
\(H_{pub}(x) = (1103515245 x + 12345) mod 327680\)


In [1]:
A = 1103515245
B = 12345
m = 327680
x = 47572947294858218452

h = (A * x + B) % m
h


283005

**Réponse 1**: \(H_{pub}(x) = 283005\)


In [2]:
import math
math.log2(327680)


18.32192809488736

**Réponse 1 (suite)**: taille ≈ 18.32 bits (≈ 18 bits).


In [3]:
# Brute force: deux collisions par h_i
A = 1103515245
B = 12345
m = 327680
h_values = [118290, 215350, 311620, 118975, 54830]

results = {h: [] for h in h_values}

x = 0
while any(len(v) < 2 for v in results.values()):
    hv = (A * x + B) % m
    if hv in results and len(results[hv]) < 2:
        results[hv].append(x)
    x += 1

results


{118290: [31133, 96669],
 215350: [6865, 72401],
 311620: [21079, 86615],
 118975: [22238, 87774],
 54830: [24745, 90281]}

**Réponse 2 (collisions)**
- h1 = 118290: x = 31133 et 96669
- h2 = 215350: x = 6865 et 72401
- h3 = 311620: x = 21079 et 86615
- h4 = 118975: x = 22238 et 87774
- h5 = 54830:  x = 24745 et 90281


**Réponse 3 : Pourquoi la fonction est cassée + collisions sans brute force ?**

H_pub(x) = (A x + B) mod m est linéaire.  
Avec A = 1103515245, B = 12345, m = 327680,  
d = gcd(A, m) = 5.                              Chaque hash a 5 préimages (si (h - B) multiple de d).  

On résout A x ≡ (h - B) (mod m) :  
- A' = A / d, m' = m / d  
- x0 = A'^{-1} * ((h - B)/d) (mod m')  
- Solutions : x = x0 + k * m' pour k = 0..d-1  

Cela redonne les collisions ci-dessus sans brute force.


## Exercice 3 — Devoir 8, exercice 6 (Diffie-Hellman)

Données:
- \(n = 1005658541636276696854926736111523240565378642222118817358603616124640001\)
- \(g = 2\)
- \(A = 694345273912301015795665084471934281997943951751938128184262171033972161\)
- \(B = 45219809272265904064518788699770061112569516384310833106022960617800893\)


**Factorisation de n** (via FactorDB):
- Factorisation de n obtenue via FactorDB (base publique), calculs cryptographiques réalisés localement.
- n = p * q
- \(p = 845722348837208875924502082373440001\)
- \(q = 1189111938473620218791973405751200001\)


**Ordres de g=2**
- ord_p = 84572234883720887592450208237344000
- ord_q = 28826956084208975001017537109120



**Pohlig–Hellman + CRT**
- a ≡ 62625926665364995110065259106721457 (mod ord_p)
- a ≡ 23795998967845128119911284363057 (mod ord_q)

Donc:
- \(a = 697755647111918244911954894837420708027739461921457\)


**Clé partagée**
- \(k = B^a mod n\)
- \(k = 922254424716744089879181468690343626534229719971081549439971098559365319\)


## Usage d'un LLM

Un modèle de langage a été utilisé pour aider à structurer les étapes de résolution, vérifier des calculs intermédiaires,
proposer un plan de solution et relire la formulation. Les idées mathématiques, la validation des résultats et la rédaction
finale ont été réalisées par l'équipe. La majorité du travail (raisonnement, choix de méthodes, validation des résultats)
est effectuée manuellement (≈70–80%).
